### ISSUE TITLE
get_zero_grad_hook device issue

### ISSUE THREAD

**Issue by jamie-mcg** [mtkresearch/FishLeg#16]
There is potentially an issue with the below function.

```
def get_zero_grad_hook(mask: torch.Tensor) -> Callable[[torch.Tensor], torch.Tensor]:
    def hook(grad: torch.Tensor) -> torch.Tensor:
        return grad * mask

    return hook
```

Basically we are getting a mismatch in the devices for `grad` and `mask` when calling `loss.backward()` in the training loop.

For now the fix below is being used:
```
def get_zero_grad_hook(mask: torch.Tensor) -> Callable[[torch.Tensor], torch.Tensor]:
    def hook(grad: torch.Tensor) -> torch.Tensor:
        mask.to(grad.get_device())
        return grad * mask

    return hook
```
but this is not very efficient.

We need to test this by running a training loop on a GPU and attempting to reproduce this error.

Also @jezabel-garcia has also discovered this issue when working in the CNN branch. She mentioned there may be a fix in there already for this.
